relayer/src/event/rpc.rs

```diff
@@ -82,19 +75,78 @@ pub fn get_all_events(
     Ok(vals)
 }
 
-fn send_packet_from_block_events(
+fn extract_block_events(
     height: Height,
-    events: HashMap<String, Vec<String>>,
+    block_events: &HashMap<String, Vec<String>>,
 ) -> Vec<(Height, IbcEvent)> {
-    let mut vals: Vec<(Height, IbcEvent)> = vec![];
-    if let Some(packets) = events.get("send_packet.packet_data") {
-        for i in 0..packets.len() {
-            let raw_obj = RawObject::new(height, "send_packet".to_string(), i, events.clone());
+    #[inline]
+    fn extract_events<'a, T: TryFrom<RawObject<'a>>>(
+        height: Height,
+        block_events: &'a HashMap<String, Vec<String>>,
+        event_type: &str,
+        event_field: &str,
+    ) -> Vec<T> {
+        block_events
+            .get(&format!("{}.{}", event_type, event_field))
+            .unwrap_or(&vec![])
+            .iter()
+            .enumerate()
+            .filter_map(|(i, _)| {
+                let raw_obj = RawObject::new(height, event_type.to_owned(), i, block_events);
+                T::try_from(raw_obj).ok()
+            })
+            .collect()
+    }
 
-            if let Ok(pkg) = ChannelEvents::SendPacket::try_from(raw_obj) {
-                vals.push((height, IbcEvent::from(pkg)))
-            }
-        }
+    #[inline]
+    fn append_events<T: Into<IbcEvent>>(
+        events: &mut Vec<(Height, IbcEvent)>,
+        chan_events: Vec<T>,
+        height: Height,
+    ) {
+        events.append(
+            &mut chan_events
+                .into_iter()
+                .map(|ev| (height, ev.into()))
+                .collect(),
+        );
     }
-    vals
+
+    let mut events: Vec<(Height, IbcEvent)> = vec![];
+    append_events::<ChannelEvents::OpenInit>(
+        &mut events,
+        extract_events(height, block_events, "channel_open_init", "channel_id"),
+        height,
+    );
+    append_events::<ChannelEvents::OpenTry>(
+        &mut events,
+        extract_events(height, block_events, "channel_open_try", "channel_id"),
+        height,
+    );
+    append_events::<ChannelEvents::OpenAck>(
+        &mut events,
+        extract_events(height, block_events, "channel_open_ack", "channel_id"),
+        height,
+    );
+    append_events::<ChannelEvents::OpenConfirm>(
+        &mut events,
+        extract_events(height, block_events, "channel_open_confirm", "channel_id"),
+        height,
+    );
+    append_events::<ChannelEvents::SendPacket>(
```

Comment:
Note: only `send_packet` events are handled. Should we handle others?

Reply:
We could but I don't see a requirement. Even for the channel events other than open_init and close_init not sure it makes sense to have. When generated in the begin/end block the implication is that the app has somehow access to the other chain, able to get proofs, etc.